README.md

# voicer

[![lifecycle](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#experimental)
[![Travis build
status](https://travis-ci.org/pmcharrison/voicer.svg?branch=master)](https://travis-ci.org/pmcharrison/voicer)
[![AppVeyor build
status](https://ci.appveyor.com/api/projects/status/github/pmcharrison/voicer?branch=master&svg=true)](https://ci.appveyor.com/project/pmcharrison/voicer)
[![Coverage
status](https://coveralls.io/repos/github/pmcharrison/voicer/badge.svg)](https://coveralls.io/r/pmcharrison/voicer?branch=master)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.2613565.svg)](https://doi.org/10.5281/zenodo.2613565)

Chord voicing is the art of assigning pitch heights to pitch classes.
This package, `voicer`, optimises the voicings of chord sequences
according to an objective cost function. The Viterbi algorithm is used
to guarantee solution optimality.

## Paper

Harrison, P. M. C., & Pearce, M. T. (2019). A computational model for
the analysis and generation of chord voicings. *PsyArXiv*.
[doi:10.31234/osf.io/wrgj7](https://doi.org/10.31234/osf.io/wrgj7)

## Installation

To install from GitHub:

``` r
if (!require(devtools)) install.packages("devtools")
devtools::install_github("pmcharrison/voicer")
```

## Example usage

`voice` is the main function to use. We pass `voice` vectors of chords
as created by the `hrep` package. Important chord representations from
the `hrep` package are:

  - `pc_set()` - create a pitch-class set (unordered pitch classes from
    0 to 11)
  - `pc_chord()` - create a pitch-class chord (a pitch-class set where
    the bass note is identified)
  - `pi_chord()` - create a pitch chord (pitches as MIDI note numbers)

<!-- end list -->

``` r
library(hrep)
library(magrittr)
library(voicer)

pc_set(c(0, 4, 7))
#> Pitch-class set: 0 4 7
pc_chord(c(4, 0, 7))
#> Pitch-class chord: [4] 0 7
pi_chord(c(60, 64, 67))
#> Pitch chord: 60 64 67
```

Using the function `vec` from the `hrep` package, we combine lists of
chords into typed vectors.

``` r
x <- list(pc_set(c(0, 4, 7)), 
          pc_set(c(0, 5, 9)),
          pc_set(c(2, 7, 11)), 
          pc_set(c(0, 4, 7))) %>% vec("pc_set")
x
#> Vector of type 'pc_set', length = 4
```

We can then pass the resulting object to `voice`.

``` r
y <- voice(x)
y 
#> Vector of type 'pi_chord', length = 4
as.list(y)
#> [[1]]
#> Pitch chord: 48 52 67 72
#> 
#> [[2]]
#> Pitch chord: 48 53 69 72
#> 
#> [[3]]
#> Pitch chord: 50 55 67 71
#> 
#> [[4]]
#> Pitch chord: 52 55 67 72
```

We can do the same for vectors of type `pc_chord`.

``` r
list(pc_chord(c(0, 4, 7)), 
     pc_chord(c(9, 0, 5)),
     pc_chord(c(7, 2, 11)), 
     pc_chord(c(0, 4, 7))) %>% 
  vec("pc_chord") %>% 
  voice() %>% 
  as.list()
#> [[1]]
#> Pitch chord: 60 64 67 72
#> 
#> [[2]]
#> Pitch chord: 57 60 65 72
#> 
#> [[3]]
#> Pitch chord: 55 62 67 71
#> 
#> [[4]]
#> Pitch chord: 60 64 67 72
```

See `?voice` for more details.


NEWS.md
# voicer 0.4.0

- Updated to reflect new version of `hrep` package.

# voicer 0.3.0

- Added `fix_melody` argument, allowing the user to prespecify a melody line.
- Added `fix_content` argument, allowing the user to mandate that the 
voicings contain specific pitch content.
- Added `fix_chords` argument, allowing the user to prespecify certain 
voicings in the chord sequence.

# voicer 0.2.0

- Added `pc_chord_type_ideal_voicings`, 
cached ideal voicings for objects of class `pc_chord_type`.

- Added a `NEWS.md` file to track changes to the package.

# voicer 0.1.0

- Initial release.


R/voice.R
#' Voice
#' 
#' Generates an optimized voicing for a chord sequence.
#' 
#' By default, the algorithm optimizes the sum of the linear predictors
#' for each chord transition.
#' By setting the \code{exp_cost}, \code{norm_cost}, and \code{log_cost}
#' arguments of \code{\link{voice_opt}} to \code{TRUE},
#' it is possible alternatively to optimize the probability of the sequence.
#' 
#' @note 
#' Voicing sequences of type \code{\link{pi_chord}} preserves 
#' the original bass pitch classes.
#' The original doublings for each pitch class can also be preserved
#' by setting \code{dbl_change = TRUE} in \code{\link{voice_opt}}.
#' 
#' @param x 
#' Chord sequence to voice.
#' This should take the form of a \code{\link[hrep]{vec}}
#' object from the \code{hrep} package.
#' A \code{\link[hrep]{vec}} object is essentially a typed list;
#' here the type should be one of three chord representations from
#' the \code{hrep} package, namely
#' \code{\link{pc_chord}}, \code{\link{pc_set}}, or \code{\link{pi_chord}}.
#' See \bold{Note} for more information.
#' 
#' @param opt
#' A list of options as created by \code{\link{voice_opt}}.
#' 
#' @param fix_melody
#' A numeric vector specifying the desired melody notes (i.e. top line)
#' for the voicing, as MIDI pitches.
#' The ith element should correspond to the melody note for the ith chord of x.
#' NA values mean that the melody is unconstrained.
#' 
#' @param fix_content
#' A list of potentially empty numeric vectors specifying pitches 
#' that must be included in the corresponding chords of x.
#' Can be used to produce melodies internal to the texture.
#' 
#' @param fix_chords
#' A list of voicings to fix in advance within the generated result,
#' each given as numeric vectors of MIDI pitches.
#' A NULL element means that the voicing for the corresponding chord is unconstrained.
#' Note that specifying a voicing manually overrides all other voicing options
#' (e.g. range limits),
#' and the voicing will be accepted even if it is incompatible
#' with the corresponding element of x.
#' 
#' @return 
#' A voiced chord sequence,
#' represented as a \code{\link[hrep]{vec}} object 
#' with type \code{\link[hrep]{pi_chord}}.
#' This object can be inspected using \code{\link[base]{as.list}}.
#' 
#' @examples
#' library(magrittr)
#' library(hrep)
#' chords <- list(
#'   pc_chord(c(0, 4, 7)),
#'   pc_chord(c(0, 5, 9)),
#'   pc_chord(c(2, 7, 11))
#' ) %>%
#'   vec("pc_chord")
#'   
#' voice(chords, 
#'       voice_opt(min_notes = 3, max_notes = 3))
#'       
#' voice(chords, 
#'       voice_opt(min_notes = 3, max_notes = 3), 
#'       fix_melody = c(76, 77, 79))
#'       
#' voice(chords, 
#'       voice_opt(min_notes = 3, max_notes = 4), 
#'       fix_content = list(c(60, 72), c(65, 77), c(67, 79)))
#'       
#' voice(chords, 
#'       voice_opt(min_notes = 3, max_notes = 3), 
#'       fix_chords = list(NULL, c(53, 57, 60), NULL))  
#'
#' @export
voice <- function(x, 
                  opt = voice_opt(), 
                  fix_melody = rep(NA_integer_, times = length(x)), 
                  fix_content = lapply(x, function(...) integer()),
                  fix_chords = vector("list", length(x))) {
  if (!(is.numeric(fix_melody) && 
        length(fix_melody) == length(x)))
    stop("fix_melody must be a numeric vector of same length as x")
  
  if (!(is.list(fix_content) && 
        length(fix_content) == length(x)))
    stop("fix_content must be a list of same length as x")
  
  if (!(is.list(fix_chords) && 
        length(fix_chords) == length(x)))
    stop("fix_chords must be a list of same length as x")
  
  UseMethod("voice")
}

#' @export
voice.vec <- function(x, 
                      opt = voice_opt(),
                      fix_melody = rep(NA_integer_, times = length(x)), 
                      fix_content = lapply(x, function(...) integer()),
                      fix_chords = vector("list", length(x))) {
  if (length(x) == 0L) return(x)
  type <- hrep::type(x)
  checkmate::qassert(type, "S1")
  f <- paste0("voice.vec_", type)
  do.call(f, args = list(x = x, 
                         opt = opt,
                         fix_melody = fix_melody, 
                         fix_content = fix_content,
                         fix_chords = fix_chords))
}

#' @export
voice.coded_vec <- function(x, 
                            opt = voice_opt(),
                            fix_melody = rep(NA_integer_, times = length(x)), 
                            fix_content = lapply(x, function(...) integer()),
                            fix_chords = vector("list", length(x))) {
  voice(hrep::decode(x), 
        opt = opt,
        fix_melody = fix_melody, 
        fix_content = fix_content,
        fix_chords = fix_chords)
}
